File: FindFirstAndLastPositionOfElementInSortedArray.java
/**
 * LeetCode 34 Find First and Last Position of Element in Sorted Array
 */
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] result = new int[]{-1,-1};
        if (nums == null || nums.length == 0) {
            return result;
        }
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int mid = (start + end) / 2;
            if (nums[mid] > target) {
                end = mid - 1;
            } else if (nums[mid] < target) {
                start = mid + 1;
            } else {
                return findStartAndEnd(nums,target,mid);
            }
        }
        return result;
    }

    public int[] findStartAndEnd(int[] nums,int target,int pos) {
        int start = -1;
        int end = -1;
        for (int i = pos; i < nums.length; i++) {
            if (nums[i] == target) {
                end = i;
            }
        }

        for (int i = pos; i >= 0; i--) {
            if (nums[i] == target) {
                start = i;
            }
        }
        return new int[]{start,end};
    }
}

File: FindFirstAndLastPositionOfElementInSortedArray.go
func searchRange(nums []int, target int) []int {
	left := binarySearch(nums,target,true)
	last := binarySearch(nums,target,false) - 1
	if (left <= last && last < len(nums) && nums[left] == target && nums[last] == target) {
		return []int{left,last}
	}
	return []int{-1,-1}
}

func binarySearch(nums []int, target int,flag bool) int {
	l := 0
	r := len(nums) - 1
	res := len(nums)

	for (l <= r) {
		mid := (l + r) / 2
		if (nums[mid] > target || (flag && nums[mid] >= target)) {
			r = mid - 1
			res = mid
		} else {
			l = mid + 1
		}
	}
	return res
}
